napari/layers/_shapes_layer/model.py

```diff
@@ -0,0 +1,1521 @@
+        self._selected_box = box + center
+
+    def _transform_box(self, transform, center=[0, 0]):
+        """Perfroms a linear transformation on the selected box
+
+        Parameters
+        ----------
+        transform : np.ndarray
+            2x2 array specifying linear transform.
+        center : list
+            coordinates of center of rotation.
+        """
+        box = self._selected_box - center
+        box = np.matmul(box, transform.T)
+        if not np.all(box[1] == box[9]):
+            transform = self.viewer._canvas.scene.node_transform(self._node)
+            rescale = transform.map([1, 1])[:2] - transform.map([0, 0])[:2]
+            r = self._rotation_handle_length*rescale.mean()
+            box[9] = box[1] + r*(box[9]-box[1])/np.linalg.norm(box[9]-box[1])
+        self._selected_box = box + center
+
+    def _shape_at(self, coord):
+        """Determines if any shapes at given coord by looking inside triangle
+        meshes.
+
+        Parameters
+        ----------
+        coord : sequence of float
+            Image coordinates to check if any shapes are at.
+
+        Returns
+        ----------
+        shape : int | None
+            Index of shape if any that is at the coordinates. Returns `None`
+            if no shape is found.
+        vertex : int | None
+            Index of vertex if any that is at the coordinates. Returns `None`
+            if no vertex is found.
+        """
+        # Check selected shapes
+        if len(self.selected_shapes) > 0:
+            if self.mode == 'select':
+                # Check if inside vertex of bounding box or rotation handle
+                inds = list(range(0, 8))
+                inds.append(9)
+                box = self._selected_box[inds]
+                distances = abs(box - coord[:2])
+
+                # Get the vertex sizes
+                scene = self.viewer._canvas.scene
+                transform = scene.node_transform(self._node)
+                rescale = transform.map([1, 1])[:2] - transform.map([0, 0])[:2]
+                sizes = self._vertex_size*rescale.mean()/2
+
+                # Check if any matching vertices
+                matches = np.all(distances <= sizes, axis=1).nonzero()
+                if len(matches[0]) > 0:
+                    return self.selected_shapes[0], matches[0][-1]
+            elif (self.mode == 'direct' or self.mode == 'vertex_insert' or
+                  self.mode == 'vertex_remove'):
+                # Check if inside vertex of shape
+                inds = np.isin(self.data._index, self.selected_shapes)
+                vertices = self.data._vertices[inds]
+                distances = abs(vertices - coord[:2])
+
+                # Get the vertex sizes
+                scene = self.viewer._canvas.scene
+                transform = scene.node_transform(self._node)
+                rescale = transform.map([1, 1])[:2] - transform.map([0, 0])[:2]
+                sizes = self._vertex_size*rescale.mean()/2
```

Comment:
I've seen this "paragraph" repeated one or two times before — good place for a function!

Reply:
fixed, good catch!